bibliopixel/control/rest.py

```diff
@@ -0,0 +1,81 @@
+import contextlib, fractions, flask, http.client, traceback, werkzeug.serving
+from . import control, editor
+from .. util import log
+from .. util.threads import runnable
+
+PORT = 8787
+
+
+class RestServer(runnable.LoopThread):
+    def __init__(self, port, external_access, get, put):
+        super().__init__()
+        self.port = port
+        self.ip_address = '0.0.0.0' if external_access else 'localhost'
+        self.app = flask.Flask(__name__)
+        self.app.route('/get/<address>')(get)
+        self.app.route('/put/<address>/<value>')(put)
+
+    def run_once(self):
+        werkzeug.serving.run_simple(self.ip_address, self.port, self.app)
+        super().stop()
+
+    def stop(self):
+        def error():
+            log.error('Unable to shut down REST server on port %d', self.port)
+
+        super().stop()
+        try:
+            flask.request.environ.get('werkzeug.server.shutdown', error)()
+        except Exception as e:
+            log.error('Exception shutting werkzeug down')
+
+
+class Rest(control.ExtractedControl):
+    """
+    REST control that uses http GET and PUT commands to get and set values
+    at Addresses in a Project.
+
+    GET commands use the URL /get/<address> and return a JSON document
+    containing the value at that address, if that exists.
+
+    PUT commands use the URL /put/<address>/<value> and then set
+    that address to be that value
+
+    In the example projects/25-rest.yml, possible URLs might be:
+
+        http://localhost:8787/get/animation.levels
+        http://localhost:8787/put/animation.levels[2]/1
+    """
+    def __init__(self, *args, port=8787, external_access=False, verbose=True,
```

Comment:
minor nitpick -- you define `PORT = 8787` in a constant up above but then use the number here. Like that it is overrideable with a default, but should either use the constant or not define it